File: iOS/YouKong/Presentation/Screens/Home/FriendScheduleViewModel.swift
import SwiftUI
import Factory

@MainActor
class FriendScheduleViewModel: ObservableObject {
    @Injected(\.agentRepository) private var agentRepository

    @Published var scheduleGroups: [ScheduleGroup] = []
    @Published var isLoading = false
    @Published var isEmpty = false

    let friendName: String
    let friendId: String

    init(friendId: String, friendName: String) {
        self.friendId = friendId
        self.friendName = friendName
    }

    func loadSchedule() async {
        isLoading = true
        defer { isLoading = false }

        do {
            let response = try await agentRepository.getUserSchedule(userId: friendId)
            let groups = response.schedules.map { ScheduleGroup(from: $0) }
            scheduleGroups = groups
            isEmpty = groups.isEmpty
        } catch {
            print("[FriendSchedule] 加载失败: \(error)")
            isEmpty = true
        }
    }

    /// 判断条目是否已执行（过去的时段）
    func isItemExecuted(_ item: ScheduleItem, in group: ScheduleGroup) -> Bool {
        guard group.isCurrentOrFuture else { return true }

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        guard let scheduleDate = formatter.date(from: group.date) else { return false }

        let calendar = Calendar.current
        let today = calendar.startOfDay(for: Date())
        let scheduleDay = calendar.startOfDay(for: scheduleDate)

        // 未来日期，未执行
        if scheduleDay > today { return false }

        // 今天：根据结束时间判断
        let now = Date()
        let timeFormatter = DateFormatter()
        timeFormatter.dateFormat = "HH:mm"
        let currentTime = timeFormatter.string(from: now)

        return item.endTime <= currentTime
    }

    /// 判断条目是否正在进行中
    func isItemActive(_ item: ScheduleItem, in group: ScheduleGroup) -> Bool {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        guard let scheduleDate = formatter.date(from: group.date) else { return false }

        let calendar = Calendar.current
        let today = calendar.startOfDay(for: Date())
        let scheduleDay = calendar.startOfDay(for: scheduleDate)

        guard scheduleDay == today else { return false }

        let now = Date()
        let timeFormatter = DateFormatter()
        timeFormatter.dateFormat = "HH:mm"
        let currentTime = timeFormatter.string(from: now)

        return currentTime >= item.startTime && currentTime < item.endTime
    }
}

File: Android/feature/feature-home/src/main/java/com/youkong/feature/home/viewmodel/FriendScheduleViewModel.kt
package com.youkong.feature.home.viewmodel

import android.util.Log
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.youkong.core.network.api.ScheduleApi
import com.youkong.core.network.model.ScheduleGroup
import com.youkong.core.network.model.ScheduleItem
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import javax.inject.Inject

data class FriendScheduleUiState(
    val scheduleGroups: List<ScheduleGroup> = emptyList(),
    val isLoading: Boolean = true,
    val isEmpty: Boolean = false,
    val friendName: String = "",
)

@HiltViewModel
class FriendScheduleViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val scheduleApi: ScheduleApi,
) : ViewModel() {

    companion object {
        private const val TAG = "FriendScheduleVM"
    }

    private val userId: String = savedStateHandle["userId"] ?: ""
    private val friendName: String = savedStateHandle["friendName"] ?: ""

    private val _uiState = MutableStateFlow(FriendScheduleUiState(friendName = friendName))
    val uiState: StateFlow<FriendScheduleUiState> = _uiState.asStateFlow()

    init {
        loadSchedule()
    }

    private fun loadSchedule() {
        if (userId.isEmpty()) {
            _uiState.update { it.copy(isLoading = false, isEmpty = true) }
            return
        }

        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }

            try {
                val response = scheduleApi.getUserSchedule(userId)
                val data = response.data

                if (data != null && data.schedules.isNotEmpty()) {
                    val groups = data.schedules.map { ScheduleGroup.fromDaySchedule(it) }
                    _uiState.update {
                        it.copy(
                            scheduleGroups = groups,
                            isEmpty = false,
                            isLoading = false
                        )
                    }
                    Log.d(TAG, "加载完成: ${groups.size} 组")
                } else {
                    _uiState.update { it.copy(isEmpty = true, isLoading = false) }
                }
            } catch (e: Exception) {
                Log.e(TAG, "加载失败: ${e.message}")
                _uiState.update { it.copy(isEmpty = true, isLoading = false) }
            }
        }
    }

    fun isItemExecuted(item: ScheduleItem, group: ScheduleGroup): Boolean {
        if (!group.isCurrentOrFuture) return true

        val today = LocalDate.now()
        val scheduleDate = try {
            LocalDate.parse(group.date, DateTimeFormatter.ISO_LOCAL_DATE)
        } catch (e: Exception) {
            return false
        }

        if (scheduleDate.isAfter(today)) return false

        // Today: check if end time has passed
        val now = LocalTime.now()
        val endTime = try {
            LocalTime.parse(item.endTime, DateTimeFormatter.ofPattern("HH:mm"))
        } catch (e: Exception) {
            return false
        }

        return now.isAfter(endTime)
    }

    fun isItemActive(item: ScheduleItem, group: ScheduleGroup): Boolean {
        val today = LocalDate.now()
        val scheduleDate = try {
            LocalDate.parse(group.date, DateTimeFormatter.ISO_LOCAL_DATE)
        } catch (e: Exception) {
            return false
        }

        if (scheduleDate != today) return false

        val now = LocalTime.now()
        val startTime = try {
            LocalTime.parse(item.startTime, DateTimeFormatter.ofPattern("HH:mm"))
        } catch (e: Exception) {
            return false
        }
        val endTime = try {
            LocalTime.parse(item.endTime, DateTimeFormatter.ofPattern("HH:mm"))
        } catch (e: Exception) {
            return false
        }

        return !now.isBefore(startTime) && now.isBefore(endTime)
    }
}
